### PR TITLE
Vmo 3604 expressions autocomplete

### PR DESCRIPTION
**Pull request by poojahpatil90**
https://user-images.githubusercontent.com/23651142/117720786-f0037780-b193-11eb-9661-da39b8a46ebb.mov

Note: I have removed some pieces of code from expressionInput.vue component as it did not add any value. Let me know if I have overlooked some pieces and have removed them.

@bzabos based on our discussion with Kyle. 
1. I believe we need to pull in date configurations from flow runner for auto complete instead of hard coding it. 
2. We might need expression auto complete in more fields of the block editor
Let me know your thoughts. I can create tickets if we decide to make these changes.

### CODE REVIEW

File: src/components/common/ExpressionInput.vue [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+      onChange: () => input.dispatchEvent(new Event('input')),
+    }, input)
+
+    this.debounceScroll = debounce(() => { this.didScroll = true }, 100)
+
+    window.addEventListener('scroll', this.debounceScroll)
+  },
+
+  beforeDestroy() {
+    window.removeEventListener('scroll', this.debounceScroll)
```
Comment (bzabos) [inline src/components/common/ExpressionInput.vue:167]: Hi @poojahpatil90 ! It doesn't look like `debouncedScroll()`'s `didScroll` effect is being consumed anywhere. Should we remove references to those altogether?

File: src/components/common/ExpressionInput.vue [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+      return methods
+    },
+    suggestions() {
+      const suggestions = [
+        ...this.contextSuggestions,
+        ...this.methodSuggestions,
+      ]
+      return [
+        ...suggestions,
+        this.topLevelSuggestions,
```
Comment (bzabos) [inline src/components/common/ExpressionInput.vue:148]: Hi @poojahpatil90, each time we use the spread (`...`) operator, we end up copying the lists. Could we avoid 1 copy by dropping them directly into the return statement's list? Not sure if that makes sense, but could we reduce this to something like:
```
return [
  ...this.contextSuggestions,
  ...this.methodSuggestions,
  ...this.topLevelSuggestions,
]
```

**Review by bzabos — APPROVED** [review]
Hi @poojahpatil90, this looks good to me! I am curious how you fixed the placement of the suggestion dropdown though 😸

File: src/components/common/ExpressionInput.vue (hunk 1 above)
Comment (poojahpatil90) [reply src/components/common/ExpressionInput.vue:167]: @bzabos You are right!  It was not being used and I have removed it's references.

File: src/components/common/ExpressionInput.vue (hunk 2 above)
Comment (poojahpatil90) [reply src/components/common/ExpressionInput.vue:148]: This makes sense!  I did not notice this part, I have optimized this part 👍

**Comment by poojahpatil90** [comment]
> Hi @poojahpatil90, this looks good to me! I am curious how you fixed the placement of the suggestion dropdown though 😸

Hi @bzabos!  There is a problem with placement when we start scrolling the block editor. It needs to be fixed and I created a ticket to address this issue - https://viamoinc.atlassian.net/browse/[VMO-4217]( https://viamoinc.atlassian.net/browse/VMO-4217)

**Comment by yeboaasare** [comment]
Hi @poojahpatil90 I observed the presence of validation errors on the Edit expression fields for a Case Block even after writing and completing the expressions. 
Screenshot 2021-06-29 at 11 20 55 AM

**Comment by poojahpatil90** [comment]
> Hi @poojahpatil90 I observed the presence of validation errors on the Edit expression fields for a Case Block even after writing and completing the expressions.
> Screenshot 2021-06-29 at 11 20 55 AM

@yeboaasare  This issue has been fixed. Please test again and let me know if you find any other issues.

**Comment by yeboaasare** [comment]
> > Hi @poojahpatil90 I observed the presence of validation errors on the Edit expression fields for a Case Block even after writing and completing the expressions.
> > Screenshot 2021-06-29 at 11 20 55 AM
> 
> @yeboaasare This issue has been fixed. Please test again and let me know if you find any other issues.



Hi @poojahpatil90 The validation errors have disappeared from the fields bit the validation notifications are still showing
Screenshot 2021-07-05 at 11 14 49 AM

**Comment by poojahpatil90** [comment]
> > > Hi @poojahpatil90 I observed the presence of validation errors on the Edit expression fields for a Case Block even after writing and completing the expressions.
> > > Screenshot 2021-06-29 at 11 20 55 AM
> > 
> > 
> > @yeboaasare This issue has been fixed. Please test again and let me know if you find any other issues.
> 
> Hi @poojahpatil90 The validation errors have disappeared from the fields but the validation notifications are still showing
> Screenshot 2021-07-05 at 11 14 49 AM

Hi @yeboaasare!   The error related to tags are planned to to be handled as part of [ VMO-4176](https://viamoinc.atlassian.net/browse/VMO-4176)

**Comment by yeboaasare** [comment]
> > > > Hi @poojahpatil90 I observed the presence of validation errors on the Edit expression fields for a Case Block even after writing and completing the expressions.
> > > > Screenshot 2021-06-29 at 11 20 55 AM
> > > 
> > > 
> > > @yeboaasare This issue has been fixed. Please test again and let me know if you find any other issues.
> > 
> > 
> > Hi @poojahpatil90 The validation errors have disappeared from the fields but the validation notifications are still showing
> > Screenshot 2021-07-05 at 11 14 49 AM
> 
> Hi @yeboaasare! The error related to tags are planned to to be handled as part of [ VMO-4176](https://viamoinc.atlassian.net/browse/VMO-4176)

Okay, thanks!

**Review by yeboaasare — APPROVED** [review]

**Pull request merged**